devdoc/waiter.md

```diff
@@ -0,0 +1,231 @@
+
+**SRS_WAITER_43_007: [** If `waiter` is `NULL`, `waiter_register_notification` shall fail and return `WAITER_RESULT_ERROR`. **]**
+
+**SRS_WAITER_43_008: [** If `notification_callback` is `NULL`, `waiter_register_notification` shall fail and return `WAITER_RESULT_ERROR`. **]**
+
+**SRS_WAITER_43_009: [** If `out_op` is `NULL`, `waiter_register_notification` shall fail and return `WAITER_RESULT_ERROR`. **]**
+
+**SRS_WAITER_43_010: [** `waiter_register_notification` shall call `interlocked_or` to set the `REGISTERED` bit and obtain the current state of the `waiter`. **]**
+
+**SRS_WAITER_43_011: [** If the `REGISTERED` bit was already set in the current state, `waiter_register_notification` shall fail and return `WAITER_RESULT_REFUSED`. **]**
+
+**SRS_WAITER_43_012: [** If the `NOTIFIED` bit is set, `waiter_register_notification` shall : **]**
+
+- **SRS_WAITER_43_042: [** call `interlocked_exchange` to increment the `EPOCH` and unset the `NOTIFIED` bit and `REGISTERED` bit. **]**
+
+- **SRS_WAITER_43_013: [** call the given `notification_callback` with `context` as the given `context`, `data` as the `data` that was given to `waiter_notify` and `result` as `WAITER_RESULT_OK`. **]**
+
+- **SRS_WAITER_43_053: [** release the reference to `data` stored in `waiter` by calling `THANDLE_ASSIGN`. **]**
+
+- **SRS_WAITER_43_014: [** call the `notify_complete_callback` given to `waiter_notify` with `context` as the `context` given to `waiter_notify` and `result` as `WAITER_RESULT_OK`. **]**
+
+- **SRS_WAITER_43_015: [** return `WAITER_RESULT_SYNC`. **]**
+
+
+**SRS_WAITER_43_016: [** If the `NOTIFIED` bit was not already set, `waiter_register_notification` shall: **]**
+
+- **SRS_WAITER_43_017: [** call `async_op_create` with `cancel_register_notification` as the cancel function. **]**
+
+- **SRS_WAITER_43_018: [** store the given `notification_callback` and the given `context` in the `waiter`. **]**
+
+- **SRS_WAITER_43_019: [** store the current state of the `waiter` in the created `THANDLE(ASYNC_OP)`. **]**
+
+- **SRS_WAITER_43_020: [** store the given `waiter` in the created `THANDLE(ASYNC_OP)` by calling `THANDLE_INITIALIZE`. **]**
+
+- **SRS_WAITER_43_021: [** store the created `THANDLE(ASYNC_OP)` in `out_op`by calling `THANDLE_INITIALIZE_MOVE`. **]**
+
+- **SRS_WAITER_43_022: [** return `WAITER_RESULT_ASYNC`. **]**
+
+**SRS_WAITER_43_023: [** If there are any failures, `waiter_register_notification` shall fail and return `WAITER_RESULT_ERROR`. **]**
+
+### cancel_register_notification
+```
+static void cancel_register_notification(void* context)
+```
+
+`cancel_register_notification` is the cancel function given to `async_op_create` when `waiter_register_notification` is called.
+
+**SRS_WAITER_43_046: [** If the current state of the `waiter` is the same as the initial state of the `waiter` when `waiter_register_notification` was called, `cancel_register_notification` shall: **]**
```

Comment:
>f the current state of the `waiter` is the same [](http://example.com/codeflow?start=26&length=47)

and if it isn't? Suppose it doesn't do anything.

Reply:
If the state has changed then nothing needs to be done.